nexus/db-queries/src/db/datastore/deployment.rs

```diff
@@ -96,6 +103,76 @@ impl DataStore {
         Self::blueprint_insert_on_connection(&conn, opctx, blueprint).await
     }
 
+    /// Creates a transaction iff the current blueprint is "bp_id".
+    ///
+    /// - The transaction is retryable and named "name"
+    /// - The "bp_id" value is checked as the first operation within the
+    /// transaction.
+    /// - If "bp_id" is still the current target, then "f" is called,
+    /// within a transactional context.
+    pub async fn transaction_if_current_blueprint_is<Func, R>(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        name: &'static str,
+        opctx: &OpContext,
+        bp_id: BlueprintUuid,
+        f: Func,
+    ) -> Result<R, Error>
+    where
+        Func: for<'t> Fn(
+                &'t async_bb8_diesel::Connection<DbConnection>,
+            ) -> Pin<
+                Box<
+                    dyn Future<Output = Result<R, TransactionError<Error>>>
+                        + Send
+                        + 't,
+                >,
+            > + Send
+            + Sync
+            + Clone,
+        R: Send + 'static,
+    {
+        let err = OptionalError::new();
+        let r = self
+            .transaction_retry_wrapper(name)
+            .transaction(&conn, |conn| {
+                let err = err.clone();
+                let f = f.clone();
+                async move {
+                    // Bail if `bp_id` is no longer the target
+                    let target =
+                        Self::blueprint_target_get_current_on_connection(
```

Comment:
I think this may need to use `SELECT ... FOR UPDATE` to prevent another transaction from changing the blueprint after we've read this but before we've finished executing `f`, assuming `f` itself doesn't change the `bp_target` table. At least, this was the conclusion I came to in trying to test this same kind of check for external networking records; there is already a helper function that takes `SelectFlavor::ForUpdate`, so I think maybe it's just changing this to match the external networking call: https://github.com/oxidecomputer/omicron/blob/28e0bd6a6422e46880ba3c01f9c6ba5e6d53e116/nexus/db-queries/src/db/datastore/deployment.rs#L859-L866

or maybe plumbing the `SelectFlavor` through `blueprint_target_get_current_on_connection` to allow that here?

Reply:
I'm on board with this change, but according to my read of https://www.cockroachlabs.com/blog/select-for-update/ , this is for performance, not correctness, right?

Also, from https://www.cockroachlabs.com/docs/stable/select-for-update , should this actually be `SELECT FOR SHARE`? The access to the blueprint target table is a read lock, not a write lock, correct?

Reply:
I guess I can't be sure that the context in which this `SELECT` will be made won't subsequently update the value, so I'll use `SELECT ... FOR UPDATE`: 78daf5d48844ffbd1c8afc7081cc6baf360d9cd6

Added a comment that this is an optimization to minimize transaction retry errors

Reply:
> I'm on board with this change, but according to my read of https://www.cockroachlabs.com/blog/select-for-update/ , this is for performance, not correctness, right?

... maybe so? I thought when I was testing this that a read of `bp_target` and another transaction changing it simultaneously allowed both to commit, but maybe I did something wrong. That memory certainly doesn't match the blog post you linked (or https://www.cockroachlabs.com/docs/stable/demo-serializable).

Reply:
Do you still have that test? If our transactions aren't actually serializable, I think an audit of basically all transactions in Omicron might be necessary

Reply:
TL;DR from our meeting on this topic:

- `SELECT FOR UPDATE` *is* necessary for update, because without it, cockroachdb can plausibly order the "read blueprint + perform update" operation before the "update blueprint" operation.
- This is documented now in 80850518ed77ec95af05ffd786810fbee5451edb

I will follow-up here by creating another issue to at least highlight where similar risks may exist elsewhere within our codebase

Reply:
Filed https://github.com/oxidecomputer/omicron/issues/6694

Reply:
okay, updating *again* thanks to discussion on https://github.com/oxidecomputer/omicron/issues/6694#issuecomment-2381010403 - `SELECT FOR UPDATE` is **not** actually necessary, though CockroachDb's rules around transaction ordering can make this somewhat non-obvious.

My TL;DR for this pattern is: "if the blueprint target **concurrently** changes **and** someone else can concurrently observe the datasets, the transaction will abort, regardless of `SELECT FOR UPDATE`. If no one concurrently observes the datasets, then the transaction **could** have happened before the blueprint got updated, and it may be logically re-ordered before the blueprint target was updated."